### PR TITLE
feature(provision): use cloud-init for ssh config

### PR DESCRIPTION
**Pull request by KnifeyMoloko**
This PR adds support for running our SCT tests with `cloud-init`-compatible `UserData`. The current implementation makes the usage conditioned on the `use_preinstalled_scylla` param being set to `false` to avoid impacting jobs that use AMIs (interfering with their `UserData`).

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/ec2_client.py
Comment (fruch) [inline sdcm/ec2_client.py:140]: if you are using Path, you can use `path.read_text()`

https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text

File: sdcm/ec2_client.py
Comment (fruch) [inline sdcm/ec2_client.py:125]: what's the idea of removing our own user data ? all the rest of the things we put into it, won't work.

Comment (KnifeyMoloko) [reply sdcm/ec2_client.py:125]: @fruch 2 points here:
1. Do we actually make any use of it though? We pass an encoded json, which EC2 `cloud_init` doesn't read (at least in artifact tests) and dumps into `user-data.txt` in `/var/lib/cloud/instances/<instance_name>`. I don't see SCT reading that file or retrieving that data using the AWS API. I've looked in the code for how we're using this data and apart from passing it in at provision time, I don't really see us using it. Am I missing something obvious?
2. If we don't replace it, we can't use `cloud-init`, since we need to pass valid `cloud-init` code in `UserData` and we're taking up that space with our custom json.

File: sdcm/ec2_client.py
Comment (KnifeyMoloko) [reply sdcm/ec2_client.py]: Good idea - fixed!

File: sdcm/ec2_client.py
Comment (vponomaryov) [reply sdcm/ec2_client.py:125]: @KnifeyMoloko User data is used outside of the SCT repo. Look at the https://github.com/scylladb/scylla-machine-image project.
And it's files:
- https://github.com/scylladb/scylla-machine-image/blob/next/common/scylla_configure.py#L85
- https://github.com/scylladb/scylla-machine-image/blob/next/common/scylla_post_start.py#L34

@fruch is right that other stuff will be broken in such a case. So, need to have a solution that won't break anything.

File: sdcm/ec2_client.py
Comment (vponomaryov) [inline sdcm/ec2_client.py:138]: We should not use relative file paths if possible.
Use `sct_abs_path` utility function [1] from SCT in such cases.

[1] https://github.com/scylladb/scylla-cluster-tests/blob/fa8df823/sdcm/sct_events/base.py#L37

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/ec2_client.py
Comment (KnifeyMoloko) [reply sdcm/ec2_client.py]: Good catch - I moved to using `sct_abs_path` here.

File: sdcm/provision/helpers/cloud_init.py
Comment (KnifeyMoloko) [reply sdcm/provision/helpers/cloud_init.py]: Fixed

File: sdcm/cluster_aws.py
Comment (soyacz) [inline sdcm/cluster_aws.py:384]: When using separate provision step in Jenkins job, this code won't be executed as initial machines provision goes through different code (https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/sct_provision/aws/cluster.py#L239).
From my understanding, this piece of code will be used only for Artifacts test which is not using provision step, so should not affect - this is just for your information.

**Review by soyacz — CHANGES_REQUESTED** [review]
I verified your run to check if this is needed (https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-just-testing/18) and from decoded user-data it looks you didn't provide this config there - just restarted sshd. Please fix and rerun test with prolonged waiting for ssh.

Besides that, in my opinion SMI should execute for artifact tests the same way as others do - can you tell why it isn't so (after you confirm it is not executed)?

File: sdcm/provision/aws-cloud-init.txt
Comment (fruch) [reply sdcm/provision/aws-cloud-init.txt:20]: Cloud init does support it

https://cloudinit.readthedocs.io/en/latest/topics/format.html#mime-multi-part-archive

It's also how I think we should implement support of multi part with our own spec in SMI

**Comment by soyacz** [comment]
Ok, now I have better understanding (I think). I'll summarize and please correct me if I'm wrong.
1. Problem with ssh keys is only for AWS as in GCE ssh settings you added is done by `startup-script` and not by `user-data` (`startup-script` is executed for each boot of VM). For Azure we use cloud-init as usual (but anyway need to add support for that in unified-provisioning)
2. For AWS we have different scenarios of VM preparation:
    1. When Scylla is installed - we use SMI. For that case we have 2 code paths - new provisioning (in separate jenkins step) and old way (which is also used during tests, when e.g. nemesis adds node)
    2. When is not installed (e.g. using official ubuntu 22 AMI) , we must configure using cloud-init

There's a way to change user-data in AWS, but this requires stopping vm. So idea could be to modify sshd configuration with cloud-init and trigger (with cloud-init too) machine stop. We'll loose all the data from ephemeral disks (but they're not used yet at this stage anyway). Then we could detect stopped state, modify user-data to SMI compatible one and start VM.
Looks cumbersome, but It's my only idea for now.

File: sdcm/cluster_aws.py
Comment (fgelcer) [reply sdcm/cluster_aws.py:384]: apart from AMI (both artifacts and upgrade) all of them use instances without Scylla pre-installed, so both will use the same mechanism...
the AMI tests do use pre-installed Scylla, so, although it for now doesn't use Ubuntu22, we still need to remember these tests exist.

**Review by fgelcer — DISMISSED** [review]
LGTM

**Comment by fgelcer** [comment]
> Ok, now I have better understanding (I think). I'll summarize and please correct me if I'm wrong.
> 
> 1. Problem with ssh keys is only for AWS as in GCE ssh settings you added is done by `startup-script` and not by `user-data` (`startup-script` is executed for each boot of VM). For Azure we use cloud-init as usual (but anyway need to add support for that in unified-provisioning)
> 2. For AWS we have different scenarios of VM preparation:
>    
>    1. When Scylla is installed - we use SMI. For that case we have 2 code paths - new provisioning (in separate jenkins step) and old way (which is also used during tests, when e.g. nemesis adds node)
>    2. When is not installed (e.g. using official ubuntu 22 AMI) , we must configure using cloud-init
> 
> There's a way to change user-data in AWS, but this requires stopping vm. So idea could be to modify sshd configuration with cloud-init and trigger (with cloud-init too) machine stop. We'll loose all the data from ephemeral disks (but they're not used yet at this stage anyway). Then we could detect stopped state, modify user-data to SMI compatible one and start VM. Looks cumbersome, but It's my only idea for now.

@soyacz , i'm not sure i understand, do you think there is a problem with the proposed solution in this PR?

**Comment by fruch** [comment]
> Ok, now I have better understanding (I think). I'll summarize and please correct me if I'm wrong.
> 
> 1. Problem with ssh keys is only for AWS as in GCE ssh settings you added is done by `startup-script` and not by `user-data` (`startup-script` is executed for each boot of VM). For Azure we use cloud-init as usual (but anyway need to add support for that in unified-provisioning)
> 2. For AWS we have different scenarios of VM preparation:
>    
>    1. When Scylla is installed - we use SMI. For that case we have 2 code paths - new provisioning (in separate jenkins step) and old way (which is also used during tests, when e.g. nemesis adds node)
>    2. When is not installed (e.g. using official ubuntu 22 AMI) , we must configure using cloud-init
> 
> There's a way to change user-data in AWS, but this requires stopping vm. So idea could be to modify sshd configuration with cloud-init and trigger (with cloud-init too) machine stop. We'll loose all the data from ephemeral disks (but they're not used yet at this stage anyway). Then we could detect stopped state, modify user-data to SMI compatible one and start VM. Looks cumbersome, but It's my only idea for now.

@soyacz you observations are correct, but I don't get the whole idea of stopping the VM, we are doing way too much at provisioning/creating the VMs

Anyhow I think for the long run this change in SMI, would help using both (like GCE  `startup-script`, cloud-init has that option too, and it would be the exact same in all clouds)
https://github.com/scylladb/scylla-machine-image/pull/363

meanwhile, I think this PR would be enough

File: sdcm/ec2_client.py
Comment (fruch) [reply sdcm/ec2_client.py:1]: I'm guessing @KnifeyMoloko has some automatic import ordering (part of him IDE)

@KnifeyMoloko in general we would prefer not doing those, it's can complicates backports, 

for example we won't backport this one, but a different PR would add import around one of those imports
and it would cause a merge conflict for cherry-picking that commit.

if you still think this ordering is a **must**, make sure it's part of the pre-commit, and then we'll fix all the code once, and keeping it like that. same as we do for autopep8. (we just did those changes, it was painful for a while, and then we can forget about it)

File: sdcm/provision/aws-cloud-init.txt
Comment (fruch) [reply sdcm/provision/aws-cloud-init.txt:20]: and here's the implementation I had in mind: https://github.com/scylladb/scylla-machine-image/pull/363

File: sdcm/ec2_client.py
Comment (KnifeyMoloko) [reply sdcm/ec2_client.py]: Yeah, I usually run the import optimization myself, this time I think I checked the automatic import optimization in PyCharm. Either way, since I'm not making changes in the code, I removed these changes to importing as well.

**Comment by KnifeyMoloko** [comment]
> Ok, now I have better understanding (I think). I'll summarize and please correct me if I'm wrong.
> 
> 1. Problem with ssh keys is only for AWS as in GCE ssh settings you added is done by `startup-script` and not by `user-data` (`startup-script` is executed for each boot of VM). For Azure we use cloud-init as usual (but anyway need to add support for that in unified-provisioning)
> 2. For AWS we have different scenarios of VM preparation:
>    
>    1. When Scylla is installed - we use SMI. For that case we have 2 code paths - new provisioning (in separate jenkins step) and old way (which is also used during tests, when e.g. nemesis adds node)
>    2. When is not installed (e.g. using official ubuntu 22 AMI) , we must configure using cloud-init
> 
> There's a way to change user-data in AWS, but this requires stopping vm. So idea could be to modify sshd configuration with cloud-init and trigger (with cloud-init too) machine stop. We'll loose all the data from ephemeral disks (but they're not used yet at this stage anyway). Then we could detect stopped state, modify user-data to SMI compatible one and start VM. Looks cumbersome, but It's my only idea for now.

I agree with @fruch - with the SMI changes we should be good. We probably don't want to make provisioning even more complex :)

**Comment by soyacz** [comment]
> @soyacz , i'm not sure i understand, do you think there is a problem with the proposed solution in this PR?

Yes, it works only with `use_preinstalled_scylla` flag. But looks it's ok for us. I understand for 5.0 we do only artifact tests with Ubuntu 22, and for next releases we'll be having updated SMI to support cloud-init.

**Comment by fgelcer** [comment]
> > @soyacz , i'm not sure i understand, do you think there is a problem with the proposed solution in this PR?
> 
> Yes, it works only with `use_preinstalled_scylla` flag. But looks it's ok for us. I understand for 5.0 we do only artifact tests with Ubuntu 22, and for next releases we'll be having updated SMI to support cloud-init.

yes, and also, we maybe will be newer keys for it, supported by the advanced version of openssh (used in Ubuntu22)

**Review by fgelcer — APPROVED** [review]
LGTM

**Review by vponomaryov — APPROVED** [review]
LGTM

**Comment by fgelcer** [comment]
@KnifeyMoloko , i have re-[run the Ubuntu22 artifacts on master](https://jenkins.scylladb.com/job/scylla-master/job/artifacts/job/artifacts-ubuntu-2204-test/5/) once this one was merged, but it seems like it is not yet passing...

**Comment by KnifeyMoloko** [comment]
@scylladb/qa-maintainers Hi Guys, please backport this to 5.0 and 2022.1 branches.

**Comment by roydahan** [comment]
Before backporting I would like to asses the risk.
In general, such PR wouldn't been backported to OSS, espcially not at this stage.
I consider it only due to 2022.1.

File: sdcm/cluster_aws.py
Comment (roydahan) [reply sdcm/cluster_aws.py:384]: Isn't this affect all OS versions and not only ubuntu 22?

**Comment by fruch** [comment]
> Before backporting I would like to asses the risk. In general, such PR wouldn't been backported to OSS, especially not at this stage. I consider it only due to 2022.1.

If we didn't encounter anything big in the weekly master, I don't think it's a big risk.

question is doesn't 5.0 or 2022.1 pending on having ubuntu22 tested, or not.

File: sdcm/cluster_aws.py
Comment (fruch) [reply sdcm/cluster_aws.py:384]: yes, but only for artifacts tests (where scylla isn't installed, and it's mostly the ARM ones, that we can't do on GCE)
and @KnifeyMoloko tested it, and proved it to be working (also on multiple OSes, that the changes to sshd config are harmless to older sshd versions)

**Comment by KnifeyMoloko** [comment]
@roydahan @scylladb/qa-maintainers  I browsed the master jobs and I don't see any issues related to this change, so I think we should be good to backport this one.

**Comment by roydahan** [comment]
It doesn't cleanly backports.
Can you please send a PR to branch-5.0 and I'll merge and backport.

**Comment by fgelcer** [comment]
@KnifeyMoloko , did you send a PR to backport this one to 5.0? if yes, please link it here

**Pull request merged**
